widgets/heatmap/heatmap.go

```diff
@@ -76,7 +88,50 @@ func New(opts ...Option) (*HeatMap, error) {
 // Each call to Values overwrites any previously provided values.
 // Provided options override values set when New() was called.
 func (hp *HeatMap) Values(xLabels []string, yLabels []string, values [][]float64, opts ...Option) error {
-	return errors.New("not implemented")
+	if len(values) == 0 {
+		return errors.New("the values cannot be empty")
+	}
+
+	hp.mu.Lock()
+	defer hp.mu.Unlock()
+
+	var (
+		xl []string
+		yl []string
+		v  [][]float64
+	)
+
+	if xLabels == nil {
+		// Set initial labels if it is not provided.
+		xl = initLabels(len(values[0]))
+	} else {
+		// Copy to avoid external modifications.
+		xl = make([]string, len(xLabels))
+		copy(xl, xLabels)
+	}
+
+	if yLabels == nil {
+		yl = initLabels(len(values))
+	} else {
+		yl = make([]string, len(yLabels))
+		copy(yl, yLabels)
+	}
+
+	if len(yl) != len(values) || len(xl) != len(values[0]) {
+		return errors.New("the number of labels does not match the shape of the values")
+	}
+
+	v = make([][]float64, len(values))
+	copy(v, values)
+
+	hp.xLabels = xl
+	hp.yLabels = yl
+	hp.values = v
+	for _, opt := range opts {
+		opt.set(hp.opts)
+	}
+	hp.minValue, hp.maxValue = minMax(hp.values)
+	return nil
 }
 
 // ClearXLabels clear the X labels.
```

Comment:
Can we ensure that all functions that change or read the state of the HeatMap do so with the mutex acquired? 

The functions `ClearXLabels` and `ClearYLabels` don't seem to be doing so currently which might lead to unexpected race conditions. Applies to all such functions.